### PR TITLE
SDK-311: Using Java 8 Optional Class

### PR DESCRIPTION
**Pull request by wikumChamith**
Jira ticket: https://issues.openmrs.org/browse/SDK-311

### CODE REVIEW

**Comment by dkayiwa** [comment]
What do you think about these changes in relation to these two articles?

http://dolszewski.com/java/java-8-optional-use-cases/

https://blogs.oracle.com/javamagazine/post/12-recipes-for-using-the-optional-class-as-its-meant-to-be-used

**Comment by wikumChamith** [comment]
> What do you think about these changes in relation to these two articles?
> 
> http://dolszewski.com/java/java-8-optional-use-cases/
> 
> https://blogs.oracle.com/javamagazine/post/12-recipes-for-using-the-optional-class-as-its-meant-to-be-used

@dkayiwa I made some changes to the code.

**Comment by dkayiwa** [comment]
Are all these changes supposed to be in this pull request?

**Comment by wikumChamith** [comment]
Ohh, I think it is a problem with rebasing. I'll recheck

**Comment by wikumChamith** [comment]
@dkayiwa  please review these new changes.

**Comment by dkayiwa** [comment]
> please review these new changes.

Feel free to continue with other pull requests as you wait for this to be reviewed. 😊

File: sdk-commons/src/main/java/org/openmrs/maven/plugins/model/DistroProperties.java [hunk 1]
```diff
@@ -122,56 +123,65 @@ public List<Artifact> getModuleArtifacts(DistroHelper distroHelper, File directo
...
+                throw new RuntimeException(e);
+            }
+        });
         return mergeArtifactLists(childArtifacts, parentArtifacts);
     }
 
     public String getPlatformVersion(DistroHelper distroHelper, File directory) throws MojoExecutionException{
         Artifact artifact = getDistroArtifact();
-        if (artifact != null) {
-            DistroProperties distroProperties = distroHelper.downloadDistroProperties(directory, artifact);
```
Comment (dkayiwa) [inline sdk-commons/src/main/java/org/openmrs/maven/plugins/model/DistroProperties.java:171]: How is this change related to the Java 8 Optional class?

File: maven-plugin/src/main/java/org/openmrs/maven/plugins/utility/OwaHelper.java [hunk 2]
```diff
@@ -524,13 +495,7 @@ public void runSystemNpmCommandWithArgs(List<String> arguments) throws MojoExecu
 	}
 
 	private String getNpmSystemExecutable() {
-		String npm;
-		if (SystemUtils.IS_OS_WINDOWS) {
```
Comment (dkayiwa) [inline maven-plugin/src/main/java/org/openmrs/maven/plugins/utility/OwaHelper.java:528]: Is this also related to the Optional class?

File: maven-plugin/src/main/java/org/openmrs/maven/plugins/utility/OwaHelper.java [hunk 3]
```diff
@@ -431,28 +413,17 @@ public SemVersion parseVersion(String version) {
...
-		}
-		return null;
+		return (packageJson != null && packageJson.getEngines() != null)? SemVersion.valueOf(packageJson.getEngines().get(NODE_VERSION_KEY)): null;
 	}
 
 	public PackageJson getPackageJson(String jsonFilename) throws MojoExecutionException {
 		Reader reader = null;
-		File json;
-		if (mavenProject == null) {
-			json = new File(jsonFilename);
```
Comment (dkayiwa) [inline maven-plugin/src/main/java/org/openmrs/maven/plugins/utility/OwaHelper.java:452]: Is this also related to the Optional class?

**Comment by dkayiwa** [comment]
@wikumChamith are all the changes in this pull request related to the Optional class?

**Comment by wikumChamith** [comment]
@dkayiwa I made the changes mentioned above according to the 12th recipe of this article you shared.

https://blogs.oracle.com/javamagazine/post/12-recipes-for-using-the-optional-class-as-its-meant-to-be-used

**Comment by wikumChamith** [comment]
> Did you see my other comments in this pull request?

Yes. I used the ternary operator because it is better than using if-else or optional ifPresent-get. It is also recommended here: https://blogs.oracle.com/javamagazine/post/12-recipes-for-using-the-optional-class-as-its-meant-to-be-used
